ops/c/src/core/ops_lazy.cpp

```diff
@@ -931,18 +939,63 @@ void ops_execute(OPS_instance *instance) {
       }
     ops_free(ops_kernel_list[i]->args);
     ops_kernel_list[i]->args = nullptr;
+    ops_free(ops_kernel_list[i]->range);
+    ops_kernel_list[i]->range = nullptr;
+    ops_free(ops_kernel_list[i]->orig_range);
+    ops_kernel_list[i]->orig_range = nullptr;
     ops_free(ops_kernel_list[i]);
     ops_kernel_list[i] = nullptr;
   }
   ops_kernel_list.clear();
 }
 
+void create_kerneldesc_and_enque(char const *name, ops_arg *args, int nargs, int index, int dim, int isdevice, int *range, ops_block block, void (*func)(struct ops_kernel_descriptor *desc))
+{
+    ops_kernel_descriptor *desc = (ops_kernel_descriptor *)calloc(1,sizeof(ops_kernel_descriptor));
+
+    desc->name = (char*) calloc(strlen(name)+1, sizeof(char));
+    strcpy(desc->name, name);       strcat(desc->name,"\0");
+    desc->name_len = strlen(name);
+    desc->block = block;
+    desc->dim = dim;
+    desc->isdevice = isdevice;
+    desc->index = index;
+    desc->hash = 5381;
+    desc->hash = ((desc->hash << 5) + desc->hash) + index;
+
+    desc->range = (int*) calloc(2*OPS_MAX_DIM, sizeof(int));
+    desc->orig_range = (int*) calloc(2*OPS_MAX_DIM, sizeof(int));
+
+    for ( int i=0; i < 2*block->dims; i++ ) {
+        desc->range[i] = range[i];
+        desc->orig_range[i] = range[i];
+        desc->hash = ((desc->hash << 5) + desc->hash) + range[i];
+    }
+
+    desc->nargs = nargs;
+    desc->args = (ops_arg*)ops_malloc(nargs*sizeof(ops_arg));
+
+    for ( int n=0; n < nargs; n++) {
+        desc->args[n] = args[n];
+
+        if (args[n].argtype == OPS_ARG_DAT)
+            desc->hash = ((desc->hash << 5) + desc->hash) + args[n].dat->index;
+        if (args[n].argtype == OPS_ARG_GBL && args[n].acc == OPS_READ) {
+            char *tmp = (char*)ops_malloc(args[n].dim*args[n].elem_size);
+            memcpy(tmp, args[n].data,args[n].dim*args[n].elem_size);
```

Comment:
is this correctly freed?

Reply:
https://github.com/OP-DSL/OPS/blob/e586f89b410eb2ed164ad9bc2d4e11ff289f0283/ops/c/src/core/ops_lazy.cpp#L935

Done it here
ops_exit_lazy(), misses freeing range and other parameter, added now